File: src/main/java/Q681_NextClosestTime.java
import java.util.HashSet;

/***
 * 25mins
 * Runtime: 8 ms, faster than 44.46% of Java online submissions for Next Closest Time.
 * Memory Usage: 39.2 MB, less than 16.88% of Java online submissions for Next Closest Time.
 */
public class Q681_NextClosestTime {
    public String nextClosestTime(String time) {
        String[] t = time.split(":");
        HashSet<Character> s = new HashSet();
        for(int i = 0; i < time.length(); i++) {
            s.add(time.charAt(i));
        }
        while(true) {
            inc(t);
            if(isValid(t, s)) {
                return t[0] + ":" + t[1];
            }
        }
    }

    public void inc(String[] time) {
        int a = Integer.valueOf(time[1]) + 1;
        int ca = 0;
        if(a == 60) {
            ca = 1;
            a = 0;
        }
        time[1] = intToString(a);
        a = Integer.valueOf(time[0]) + ca;
        if(a == 24) {
            a = 0;
        }
        time[0] = intToString(a);

    }

    public String intToString(int a) {
        if(a < 10) {
            return "0" + a;
        }
        else {
            return "" + a;
        }
    }

    public boolean isValid(String[] time, HashSet<Character> s) {
        String x = "" + time[0] + time[1];
        for(int i = 0; i < x.length(); i++) {
            if(!s.contains(x.charAt(i))) {
                return false;
            }
        }
        return true;
    }
}

File: LeetCodePractice/LeetCodePractice/Q681_NextClosestTime.cpp
#include "LCHeader.h"
//8:56 9:09
/*
Runtime: 0 ms, faster than 100.00% of C++ online submissions for Next Closest Time.
Memory Usage: 6 MB, less than 54.76% of C++ online submissions for Next Closest Time.
*/

class Solution {
public:
    string nextClosestTime(string time) {
        unordered_set<char> set(time.cbegin(), time.cend());
        next(time);
        while (true) {
            bool pass = true;
            for (auto& x : time) {
                if (set.count(x) == 0) {
                    pass = false;
                }
            }
            if (pass) {
                return time;
            }
            next(time);
        }
    }

    void next(string& time) {
        if (time[4] < '9') {
            time[4]++;
            return;
        }
        if (time[4] == '9') {
            time[3]++;
            time[4] = '0';
        }
        bool add = false;
        if (time[3] == '6') {
            time[3] = '0';
            add = true;
        }
        if (!add) {
            return;
        }
        if (time[0] < '2') {
            if (time[1] == '9') {
                time[0]++;
                time[1] = '0';
            }
            else {
                time[1]++;
            }
        }
        else {
            if (time[1] == '3') {
                time[0] = '0';
                time[1] = '0';
            }
            else {
                time[1]++;
            }
        }
    }

};